src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecutionTest.java

```diff
@@ -477,6 +477,33 @@ private void trustedShell(final boolean pos) {
         });
     }
 
+    @Test
+    public void existingVariablesOnRestart() throws Exception {
+        story.then(r -> {
+            WorkflowJob p = r.jenkins.createProject(WorkflowJob.class, "p");
+            p.setDefinition(new CpsFlowDefinition("echo('pre-semaphore')\n" +
```

Comment:
I would update the `echo` steps to have them use the injected variable, and then check the logs and assert that it was present both before and after the restart.

```suggestion
            p.setDefinition(new CpsFlowDefinition("echo('Pre-semaphore value is ${existing}')\n" +
```

And then you can add an assertion like this before the restart, and a similar one afterwards: 
```
r.assertLogContains("Pre-semaphore value is value", b);
```

Reply:
addressed in f15b02ddf3268eac1df80d25f14745975d68baa4